csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -101,10 +101,11 @@ public void AddEntriesFrom(CodedInputStream input, FieldCodec<T> codec)
                 int length = input.ReadLength();
                 if (length > 0)
                 {
+                    EnsureSize(count + length);
```

Comment:
You can still do this optimization for fixed-width field types such as "fixed32", "fixed64", "float", "double". For those you know for sure the number of elements is length/4 or length/8. I don't really know much about the C# implementation yet, but if FieldCodec has some methods to indicate whether a field is fixed-length or not, it should be easy to add an optimized code path for fixed-length field types here.

And for varint, it might not hurt to make an guess on the average length of a single varint, and use it to call EnsureSize(). We need more data for this guess work though.